### PR TITLE
Enable address sanitizer in rocm's llvm-amdgpu package.

### PR DESCRIPTION
**Pull request by srekolam**
this PR will enable the address sanitizer changes inside llvm-amdgpu package. 
each of the individual rocm packages will inturn need additional changes which will be done as part of the rocm-5.7.0 updates.
PR #40367 mistakenly removes def cmake_args(self) inside hsakmt-roct package which is restored now.

### CODE REVIEW

**Review by renjithravindrankannath — APPROVED** [review]

**Comment by srekolam** [comment]
@tldahlgren , can you please review and let me know if this can be merged.

**Comment by srekolam** [comment]
@tldahlgren , can you please let us know if you can review the changes and merge if this is okay.

**Comment by srekolam** [comment]
@tldahlgren  can you please review this PR. Thanks.

File: var/spack/repos/builtin/packages/llvm-amdgpu/package.py [hunk 1]
```diff
@@ -162,6 +162,7 @@ class LlvmAmdgpu(CMakePackage):
     # as per 5.2.0 llvm code. It used to be llvm/bin/../lib/libdevice.
     # Below patch is to look in the old path.
     patch("adjust-openmp-bitcode-directory-for-llvm-link.patch", when="@5.2.0:")
+    patch("adjust-openmp-bitcode-directory-for-llvm-link.patch", when="@5.2.0:5.6")
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/llvm-amdgpu/package.py]: Did you mean to use only one patch directive here?

File: var/spack/repos/builtin/packages/llvm-amdgpu/package.py [hunk 2]
```diff
@@ -227,6 +228,48 @@ class LlvmAmdgpu(CMakePackage):
...
+    for d_version, d_shasum in [
+        ("5.6.1", "4de9a57c2092edf9398d671c8a2c60626eb7daf358caf710da70d9c105490221"),
+        ("5.6.0", "30875d440df9d8481ffb24d87755eae20a0efc1114849a72619ea954f1e9206c"),
+    ]:
+        resource(
+            name="hsa-runtime",
+            placement="hsa-runtime",
+            url="https://github.com/RadeonOpenCompute/ROCR-Runtime/archive/rocm-{0}.tar.gz".format(
+                d_version
+            ),
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/llvm-amdgpu/package.py]: FWIW. We're encouraging use of f-strings in the core so you could switch this here while you're making the change:

```suggestion
            url=g"https://github.com/RadeonOpenCompute/ROCR-Runtime/archive/rocm-{d_version}.tar.gz"
```

**Review by tldahlgren — COMMENTED** [review]
Looking good.  I have a question about the addition (versus modification) of a patch directive and a recommendation to start switching to f-strings.

File: var/spack/repos/builtin/packages/llvm-amdgpu/package.py (hunk 1 above)
Comment (srekolam) [reply var/spack/repos/builtin/packages/llvm-amdgpu/package.py]: it is corrected now.

File: var/spack/repos/builtin/packages/llvm-amdgpu/package.py [hunk 3]
```diff
@@ -235,9 +234,7 @@ class LlvmAmdgpu(CMakePackage):
         resource(
             name="hsa-runtime",
             placement="hsa-runtime",
-            url="https://github.com/RadeonOpenCompute/ROCR-Runtime/archive/rocm-{0}.tar.gz".format(
-                d_version
-            ),
+            url="https://github.com/RadeonOpenCompute/ROCR-Runtime/archive/rocm-{d_version}.tar.gz",
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/llvm-amdgpu/package.py]: Apologies if I forgot to include the `f` in my suggestions:

```suggestion
            url=f"https://github.com/RadeonOpenCompute/ROCR-Runtime/archive/rocm-{d_version}.tar.gz",
```

File: var/spack/repos/builtin/packages/llvm-amdgpu/package.py [hunk 4]
```diff
@@ -256,9 +253,7 @@ class LlvmAmdgpu(CMakePackage):
         resource(
             name="comgr",
             placement="comgr",
-            url="https://github.com/RadeonOpenCompute/ROCm-CompilerSupport/archive/rocm-{0}.tar.gz".format(
-                d_version
-            ),
+            url="https://github.com/RadeonOpenCompute/ROCm-CompilerSupport/archive/rocm-{d_version}.tar.gz",
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/llvm-amdgpu/package.py]: Same here. 🙍‍♀️ 

```suggestion
            url=f"https://github.com/RadeonOpenCompute/ROCm-CompilerSupport/archive/rocm-{d_version}.tar.gz",
```

**Review by tldahlgren — CHANGES_REQUESTED** [review]
Apologies for forgetting to include the required `f` in the f-strings.

**Review by tldahlgren — APPROVED** [review]

**Pull request merged**